src/PowerShellEditorServices/Language/FindDeclarationVisitor.cs

```diff
@@ -76,19 +78,68 @@ public override AstVisitAction VisitFunctionDefinition(FunctionDefinitionAst fun
         /// or a decision to continue if it wasn't found</returns>
         public override AstVisitAction VisitAssignmentStatement(AssignmentStatementAst assignmentStatementAst)
         {
-            var variableExprAst = assignmentStatementAst.Left as VariableExpressionAst;
-            if (variableExprAst == null ||
-                variableName == null ||
-                !variableExprAst.VariablePath.UserPath.Equals(
+            if (variableName == null)
+            {
+                return AstVisitAction.Continue;
+            }
+
+            // The AssignmentStatementAst could contain either of the following Ast types:
+            // VariableExpressionAst, ArrayLiteralAst, ConvertExpressionAst, AttributedExpressionAst
+            // We might need to recurse down the tree to find the VariableExpressionAst we're looking for
+            List<VariableExpressionAst> asts = FindMatchingAsts(assignmentStatementAst.Left, variableName);
+            if (asts.Count > 0)
+            {
+                FoundDeclaration = new SymbolReference(SymbolType.Variable, asts.FirstOrDefault().Extent);
+                return AstVisitAction.StopVisit;
+            }
+            return AstVisitAction.Continue;
+        }
+
+        /// <summary>
+        /// Takes in an ExpressionAst and recurses until it finds all VariableExpressionAst and returns the list of them.
+        /// We expect this ExpressionAst to be either: VariableExpressionAst, ArrayLiteralAst, ConvertExpressionAst, AttributedExpressionAst
+        /// </summary>
+        /// <param name="ast">An ExpressionAst</param>
+        /// <param name="variableName">The name of the variable we are trying to find</param>
+        /// <returns>A list of ExpressionAsts that match the variable name provided</returns>
+        private static List<VariableExpressionAst> FindMatchingAsts (ExpressionAst ast, string variableName) {
+
+            // VaraibleExpressionAst case - aka base case. This will return a list with the variableExpressionAst in it or an empty list if it's not the right one.
+            var variableExprAst = ast as VariableExpressionAst;
+            if (variableExprAst != null)
```

Comment:
Since the variable will only be used within the `if` block, an is assignment could be used here instead. (e.g. `if (ast is VariableExpressionAst variableExprAst) {}`)

Or possibly an assignment switch

```csharp
switch (ast)
{
    case ConvertExpressionAst convertExpression:
    {
        FindMatchingAsts(convertExpressionAst);
        break;
    }
}
```

Reply:
Yeah I reckon a type switch is probably clearest here -- it's probably how the PowerShell compiler itself would work if it were written today.

Reply:
I saw a bunch of articles that said this wasn't possible hmm - I'll give it a go

Reply:
The MS Docs seem to call it [pattern matching](https://docs.microsoft.com/en-us/dotnet/csharp/pattern-matching). (Which is guess is approximately true...)

Reply:
Ohhhh, but this may not be possible in all the .NET versions we support

Reply:
(Historical note: We think this should work)